### PR TITLE
Fmegen/adding crl inmemory caching

### PR DESCRIPTION
**Pull request by fmegen**

### CODE REVIEW

**Review by zhouwangzw — APPROVED** [review]

File: adapters/tlsio_openssl.c [hunk 1]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+static int crl_cache_size = 0;
+static X509_CRL** crl_cache = NULL;
+static int load_cert_crl_memory(X509 *cert, X509_CRL **pCrl)
+{
+    time_t now = time(NULL);
+    X509_NAME *issuer_cert = X509_get_issuer_name(cert);
+
+    // init return values
+    int ret = 0;
+    *pCrl = NULL;
```
Comment (zhouwangzw) [inline adapters/tlsio_openssl.c:973]: pCrl could be null?

File: adapters/tlsio_openssl.c [hunk 2]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+
+    return result;
+}
+
+static time_t crl_invalid_after(X509_CRL *crl)
+{
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+    const ASN1_TIME *at = X509_CRL_get0_nextUpdate(crl);
+#else
+    ASN1_TIME *at = crl->crl->nextUpdate;
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:920]: crl and crl->crl cannot be null, right?

File: adapters/tlsio_openssl.c [hunk 3]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+#endif
+
+    ASN1_GENERALIZEDTIME *gt = ASN1_TIME_to_generalizedtime(at, NULL);
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:923]: what if "at" is null? does ANS1 handles this gracefully?

File: adapters/tlsio_openssl.c [hunk 4]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
+    {
+        return 0;
+    }
+
+    // Important: At this point, we will DELETE
+    //      a file holding a Crl from disk in case
+    //      the invalid-after date is less than the
+    //      current time.
+    //      This will trigger the re-loading of the
+    //      Crl from the download store, if available.
```
Comment (zhouwangzw) [inline adapters/tlsio_openssl.c:1295]: Is this comment still valid?

File: adapters/tlsio_openssl.c [hunk 5]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) && (OPENSSL_VERSION_NUMBER < 0x20000000L)
+    const ASN1_TIME *at = X509_CRL_get0_nextUpdate(crl);
+#else
+    ASN1_TIME *at = crl->crl->nextUpdate;
+#endif
+
+    ASN1_GENERALIZEDTIME *gt = ASN1_TIME_to_generalizedtime(at, NULL);
+    if (!gt)
+    {
+        LogError("crl could not find field\n");
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:926]: which field?

File: adapters/tlsio_openssl.c [hunk 6]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+        return 0;
+    }
+    // LogInfo("crl data %s\n", (char*)gt->data);
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:929]: remove

File: adapters/tlsio_openssl.c [hunk 7]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+        tm.tm_year = atoin((char*)gt->data, 0, 4) - 1900;
+        tm.tm_mon = atoin((char*)gt->data, 4, 2) - 1;
+        tm.tm_mday = atoin((char*)gt->data, 6, 2);
+        tm.tm_hour = atoin((char*)gt->data, 8, 2);
+        tm.tm_min = atoin((char*)gt->data, 10, 2);
+        tm.tm_sec = atoin((char*)gt->data, 12, 2);
+
+        success = (tm.tm_year > 100) &&
+            (tm.tm_mon >= 0 && tm.tm_mon  < 12) &&
+            (tm.tm_mday > 0 && tm.tm_mday < 32) &&
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:946]: you are sure that days differently from months and other times start with 1, right?:) weird :)

File: adapters/tlsio_openssl.c [hunk 8]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+
+    // init return values
+    int ret = 0;
+    *pCrl = NULL;
+
+    LOCK_RESULT lockResult = Lock(crl_cache_lock);
+    if (LOCK_OK != lockResult)
+    {
+        // could not lock
+        return 0;
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:979]: not sure whether 0 indicates success here?

File: adapters/tlsio_openssl.c [hunk 9]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+        *pCrl = crl;
+        ret = 1;
+        break;
+    }
+
+    lockResult = Unlock(crl_cache_lock);
+    return ret;
+}
+
+static int save_cert_crl_memory(X509 *cert, X509_CRL *crlp)
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1022]: could you please comment a bit more what the functions are supposed to do?

File: adapters/tlsio_openssl.c [hunk 10]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+            crl_cache[n] = crlp;
+
+            lockResult = Unlock(crl_cache_lock);
+            return 1;
+        }
+    }
+
+    // allocate bigger array
+    X509_CRL** new_crl_cache;
+    new_crl_cache = (X509_CRL**)malloc((crl_cache_size + 10) * sizeof(X509_CRL*));
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1094]: cant you use realloc?

File: adapters/tlsio_openssl.c [hunk 11]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+    if (!new_crl_cache)
+    {
+        lockResult = Unlock(crl_cache_lock);
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1097]: you are leaking crlp in this return happens

Comment (eldakms) [reply adapters/tlsio_openssl.c:1097]: "if this return happens", sorry :)

File: adapters/tlsio_openssl.c [hunk 12]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+    if (!success)
+    {
+        return 0;
+    }
+
+    // calculates the time since epoch
+    return mktime(&tm);
+}
+
+
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:962]: could you please prefix with g_, so that it is clear that they are global.

File: adapters/tlsio_openssl.c [hunk 13]
```diff
@@ -897,6 +897,226 @@ static int load_cert_crl_http(
...
+    memcpy(new_crl_cache, crl_cache, crl_cache_size * sizeof(X509_CRL*));
+    memset(new_crl_cache + crl_cache_size, 0, 10 * sizeof(X509_CRL*));
+    new_crl_cache[crl_cache_size] = crlp;
+
+    // finally update the cache
+    X509_CRL** old_crl_cache = crl_cache;
+    crl_cache = new_crl_cache;
+    crl_cache_size += 10;
+
+    if (old_crl_cache) {
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1111]: formatting

File: adapters/tlsio_openssl.c [hunk 14]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
+    if (0 != X509_NAME_cmp(issuer_crl, issuer_cert))
+    {
+        return 0;
+    }
+
+    // Important: At this point, we will DELETE
+    //      a file holding a Crl from disk in case
+    //      the invalid-after date is less than the
+    //      current time.
+    //      This will trigger the re-loading of the
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1294]: where is this delete happening? is the comment outdated?

File: adapters/tlsio_openssl.c [hunk 15]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
+static int is_valid_crl(X509 *cert, X509_CRL *crl)
 {
-    int i;
-    X509_CRL *crl = NULL;
-    char buf[256];
     time_t now = time(NULL);
+    X509_NAME *issuer_cert = X509_get_issuer_name(cert);
+
+    // names don't match up. probably a hash collision
+    // so lets test if there is another crl on disk.
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1283]: not sure what the comment means, where do we get anything from disk?

File: adapters/tlsio_openssl.c [hunk 16]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
 
     // we need the issuer hash to find the file on disk
     X509_NAME *issuer_cert = X509_get_issuer_name(cert);
     unsigned long hash = X509_NAME_hash(issuer_cert);
 
     // try to read from file
-    for (i = 0; i < 10; i++)
+    for (int i = 0; i < 10; i++)
     {
         sprintf(buf, "%s/%08lx.%s.%d", prefix, hash, suffix, i);
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1328]: < 256... can we use some functions that check the boundary?

File: adapters/tlsio_openssl.c [hunk 17]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
+        return 0;
     }
 
     // we need the issuer hash to find the file on disk
     X509_NAME *issuer_cert = X509_get_issuer_name(cert);
     unsigned long hash = X509_NAME_hash(issuer_cert);
 
     // try to read from file
-    for (i = 0; i < 10; i++)
+    for (int i = 0; i < 10; i++)
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1326]: not sure where 10 is coming from?

File: adapters/tlsio_openssl.c [hunk 18]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
-        //      the invalid-after date is less than the
-        //      current time.
-        //      This will trigger the re-loading of the
-        //      Crl from the download store, if available.
-        time_t crlend = crl_invalid_after(crl);
-        if (crlend <= now)
-        {
-            LogInfo("crl %ld, %ld DELETE %s\n", crlend, now, buf);
 #ifdef WIN32
             _unlink(buf);
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1343]: mmm? unlink char[]?

File: adapters/tlsio_openssl.c [hunk 19]
```diff
@@ -1116,57 +1274,71 @@ static const char *get_dp_url(DIST_POINT *dp)
...
         if (!crl)
         {
             continue;
         }
 
-        // names don't match up. probably a hash collision
-        // so lets test if there is another crl on disk.
-        X509_NAME *issuer_crl = X509_CRL_get_issuer(crl);
-        if (0 != X509_NAME_cmp(issuer_crl, issuer_cert))
+        if (!is_valid_crl(cert, crl))
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1338]: aren't you leaking here crl? I mean when you load it you do reference++, I do not see where you do --/refdecrease.

File: adapters/tlsio_openssl.c [hunk 20]
```diff
@@ -1177,7 +1349,64 @@ static X509_CRL *load_crl_crldp(X509 *cert, const char* suffix, STACK_OF(DIST_PO
...
+    {
+        return 0;
+    }
+
+    // we need the issuer hash to find the file on disk
+    X509_NAME *issuer_cert = X509_get_issuer_name(cert);
+    unsigned long hash = X509_NAME_hash(issuer_cert);
+
+    for (int i = 0; crl && i < 10; i++)
+    {
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1377]: why 10 times?

File: adapters/tlsio_openssl.c [hunk 21]
```diff
@@ -1673,6 +1894,8 @@ static int create_openssl_instance(TLS_IO_INSTANCE* tlsInstance)
 
 int tlsio_openssl_init(void)
 {
+    crl_cache_lock = Lock_Init();
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1897]: No uninit?

File: adapters/tlsio_openssl.c [hunk 22]
```diff
@@ -1177,7 +1349,64 @@ static X509_CRL *load_crl_crldp(X509 *cert, const char* suffix, STACK_OF(DIST_PO
...
 
-        // at this point, we got a valid crl
+        *pCrl = crl;
+        ret = 1;
+    }
+
+    return ret;
+}
+
+static int save_cert_crl_file(X509 *cert, const char* suffix, X509_CRL *crl)
```
Comment (eldakms) [inline adapters/tlsio_openssl.c:1359]: in many functions you return error code, but these are ignored on the calling side in several places. i'm not sure why you need then to return anything...

**Review by eldakms — COMMENTED** [review]
I would prefer if you submit a CodeFlow PR instead of github...

File: adapters/tlsio_openssl.c (hunk 3 above)
Comment (fmegen) [reply adapters/tlsio_openssl.c:923]: I would hope so

File: adapters/tlsio_openssl.c (hunk 5 above)
Comment (fmegen) [reply adapters/tlsio_openssl.c:926]: nextUpdate

File: adapters/tlsio_openssl.c (hunk 7 above)
Comment (fmegen) [reply adapters/tlsio_openssl.c:946]: yes, this is weird. (I double checked the docs.)

File: adapters/tlsio_openssl.c (hunk 18 above)
Comment (fmegen) [reply adapters/tlsio_openssl.c:1343]: since I created the filename, i know for sure there is no utf8 in it ;)

**Pull request merged**